File: Capítulo_10/Árvore binária e árvore binária de busca/Pesquisando valores em  árvore/Pesquisando valores mín e máx/Código em Python/max.py
# Compare elemets
Compare = {
    "LESS_THAN": -1,
    "BIGGER_THAN": 1
}

# Função default_compare
def default_compare(a, b):
    if (a == b):
        return 0
    
    return Compare["LESS_THAN"] if a < b else Compare["BIGGER_THAN"]

# Classe Node
class Node:
    def __init__(self, key):
        self.key = key # valor do nó
        self.left = None # refêrencia ao nó filho à esquerda
        self.right = None # refêrencia ao nó filho à esquerda

# Classe BinarySeachTree:
class BinarySearchTree:
    def __init__(self, compare_fn = default_compare):
        self.compare_fn = compare_fn # usado para comparar os valores dos nós
        self.root = None # nó raiz do tipo Node

    # max(): esse método devolve a chave/valor máximo da árvore.
    def max(self):
        return self.max_node(self.root)
    
    # Método auxiliar max_node(node):
    def max_node(self, node):
        current = node
        while current and current.right is not None:
            current = current.right
        return current 

File: Capítulo_10/Árvore binária e árvore binária de busca/Pesquisando valores em uma árvore/Pesquisando valores mínimos e máximos/max.js
import { Compare, defaultCompare } from "../../../models/util.js";
import { Node } from "../../../models/node.js"

export default class BinarySearchTree {
    constructor(compareFn = defaultCompare) {
        this.compareFn = compareFn; // usado para comparar os valores dos nós
        this.root = null; // nó raiz do tipo Node
    }

    // max(): esse método devolve a chave/valor máximo da árvore.
    max() {
        return this.maxNode(this.root)
    }

    // Método auxiliar maxNode(node):
    maxNode(node) {
        let current = node;
        while (current != null && current.right != null) {
            current = current.right;
        }
        return current;
    }
}